File: ai-service/app/normalize.py
import re
import unicodedata


def normalize_text(text: str) -> str:
    text = unicodedata.normalize("NFC", text)
    text = re.sub(r"[^가-힣a-zA-Z0-9\s]", "", text)
    text = re.sub(r"\s+", " ", text)
    return text.strip()

File: frontend/src/shared/utils/normalize.ts
const CLEAN_PATTERN = /[^가-힣a-zA-Z0-9\s]/g;
const COLLAPSE_PATTERN = /\s+/g;

/**
 * BE TextNormalizer.normalize()와 동일한 로직.
 * NFC 정규화 → 특수문자 제거 → 공백 정리.
 */
export function normalizeGuessText(text: string): string {
  const nfc = text.normalize("NFC");
  const cleaned = nfc.replace(CLEAN_PATTERN, "");
  const collapsed = cleaned.replace(COLLAPSE_PATTERN, " ");
  return collapsed.trim();
}
